### DISCUSSION TITLE
Trigger servo motor and optogenetic using camera input

### DISCUSSION THREAD

**Discussion by salvatoreNRIM** [bonsai-rx/bonsai#1386] [Q&A]
Hi all,

I already have a bonsai workflow (thanks to bruno's help) that tracks the centroid from a camera to trigger optogenetic stimulation using an arduino script. Now I was trying to use the same node used to trigger a servo motor as well. However I run into the error 'No coercion operator is defined between types 'System.String' and 'System.Int32'.] Would anyone have some suggestions on how to solve this? 
[trigger_opto_press_S_SL.zip](https://github.com/bonsai-rx/bonsai/files/14402220/trigger_opto_press_S_SL.zip)

![image](https://github.com/bonsai-rx/bonsai/assets/46469676/62b0f377-7ae7-4475-8b58-10960b302642)

Thanks!
Salvo

**Comment by bruno-f-cruz** [comment]
You need to convert the string to a int since the ServouOutput expects a numeric value.

**Comment by salvatoreNRIM** [comment]
Hi Bruno,

Sorry for the confusion... I use Bonsai to trigger the optogenetic stimulation via the arduino script attached

[OptoArduinoSerial.ino.zip](https://github.com/bonsai-rx/bonsai/files/14395702/OptoArduinoSerial.ino.zip)

**Comment by bruno-f-cruz** [comment]
There is an important point to make here. In order to use the Arduino package, you must flash your arduino board with Firmata. This tutorial might clear some things up (https://github.com/bruno-f-cruz/docs-wip/blob/%2320/articles/ArduinoArticle.md).
Since you are using your own custom script, you will not be able to use the arduino package at all. If you want to use a servo, you will need to come up with your own protocol logic (e.g. seding the command `m 45` will move the servo motor by 45 degrees) and implement the servo control in the arduino script using the appropriate library (probably `Servo`).

Hope this helps,
Bruno

**Comment by salvatoreNRIM** [comment]
Hi Bruno, 
Thanks for the link, indeed I always upload the Firmata before starting the workflow. And you were right, I uploaded the wrong Bonsai's file, I edited my original question with the correct file. Sorry about the confusion.
Just to recap what I intend to do with this workflow is detecting an object, and when the object is in a certain (randomized) position, if I pressed S on my keyboard beforehand, some arduino script will be initiated. I will double check the arduino script to see whether that solves the issue.
However I still have one issue on the Bonsai's side: the arduino script is initiated when the object is detected regardless of whether I pressed S or not, while I would want this stream of the workflow to be initiated conditional that I pressed S
![image](https://github.com/bonsai-rx/bonsai/assets/46469676/5c2b5fca-ed46-48a7-9baf-dde5388e07b3)

**Comment by bruno-f-cruz** [comment]
Hi,

Sorry I am probably missing something very obvious, but the script you have in the original post (the one you edited) it is just sending a character 's' to a generic serial device. There is no reference to an Arduino nor to the Arduino package whatsoever. I am very confused when you say 'I always upload Firmwata before starting the workflow'. Uploading firmata would put the Arduino in a state where sending that character would not even trigger any behavior.

**Comment by salvatoreNRIM** [comment]
Hi Bruno,

Not sure that answers your question, but the way that I communicate with the arduino is through the serial port node on the top right corner, below the details:
![image](https://github.com/bonsai-rx/bonsai/assets/46469676/c04cd303-868f-413a-9feb-40c0a1748a35)

Then the SerialStringWrite node should send the string to the arduino. Before I start the workflow I upload the StandardFirmata to flush out previous scripts, then I upload the arduino script I have above (OptoArduinoSerial.ino), then I start the workflow.
The behavior I observe at this point is that the LED gets triggered at random point when the stimulus is visible (so there is for sure some communication from Bonsai to Arduino). The issue that I have is that this communication is not conditional to arduino receiving a string (which is the behavior I would expect from the workflow above), but I am probably missing something...

**Reply by bruno-f-cruz** [reply]
Uploading firmata is unnecessary. When you upload your script it already "flushes" whatever script was in the arduino previously. 

How are you validating that:
1. "S" is being sent when it should from bonsai
2. What is triggering the laser is the s being received by the arduino and not some circuit related issue?

**Comment by salvatoreNRIM** [comment]
Hi @bruno-f-cruz 

Sorry for late reply on this. and I appreciate the clarification

1. I am not sure how to verify this, but the laser is triggered only when the bonsai script is running, and when the object is visible under the camera (and at random positions). So I would say that the bonsai script is running as I would expect, except for the fact that pressing the S does not act as a "gate" to trigger the stimulation at specific times
2. If I understand this correctly, similarly to 1, the laser is triggered only when the bonsai script is running.

Please let me know if you have any other questions

**Reply by bruno-f-cruz** [reply]
If you open the SerialWrite operator while bonsai is running, does the S string get emitted when you expect to see a trigger given your control logic?

**Comment by salvatoreNRIM** [comment]
It does. I actually have misunderstood the logic of your script - I thought that the trigger would be conditioned to press the S key. I now realize this was just a way to create communication with the Arduino... Really sorry about that 

To get back to my original question. I would want to use this script to trigger both the optogenetic stimulation and a servo motor (at the same time). To do so I adapted the script in the post as attached 
[Opto_Servo_combination_March2024.zip](https://github.com/bonsai-rx/bonsai/files/14648665/Opto_Servo_combination_March2024.zip)

However, when I run this script, I run into the issue: no COM4 port found, which I guess is because those two nodes are conflicting with each other
![image](https://github.com/bonsai-rx/bonsai/assets/46469676/c7df2593-2920-4edf-b512-5847cd9cadeb)

I tested the two streams after the condition node separately, and they respectively activate the opto trigger:
![image](https://github.com/bonsai-rx/bonsai/assets/46469676/22861d2b-4b68-4c11-a8d3-71adf8bba029)

And the servo motor:
![image](https://github.com/bonsai-rx/bonsai/assets/46469676/d1f8c394-d896-424c-9b00-d09b3bfe3619)

However, the issue is when combining the two together. I am probably missing some obvious way to integrate the two...
